File: src/java/org/krvoje/bibd/IncidenceStructure.java
package org.krvoje.bibd;

public interface IncidenceStructure {

    public int v();
    public int r();
    public int b();
    public int k();
    public int lambda();

    public int heuristicDistance();

    public boolean isBIBD();
    public void flip(int row, int col);
    public boolean active(int row, int col);
    public boolean dormant(int row, int col);

    public int incidences(int row, int col);
    public void setIncidence(int row, int col, boolean active);

    public int rowIntersection(int row, int otherRow);
    public int colIntersection(int col, int otherCol);

    public int sumInRow(int row);
    public int sumInCol(int col);

}

File: src/main/scala/org/krvoje/bibd/IncidenceStructure.scala
package org.krvoje.bibd

trait IncidenceStructure {

  val v: Int
  val r: Int
  val b: Int
  val k: Int
  val lambda: Int

  def heuristicDistance: Int
  val maxHeuristicDistance = b * (v - k) + (v * v - v) * (r - lambda)

  def isBIBD: Boolean

  def flip(row: Int, Col: Int): Unit
  def active(row: Int, Col: Int): Boolean
  def dormant(row: Int, Col: Int): Boolean

  def incidences(row: Int, Col: Int): Int
  def setIncidence(row: Int, Col: Int, active: Boolean): Unit

  def rowIntersection(row: Int, Col: Int): Int
  def colIntersection(row: Int, Col: Int): Int

  def sumInRow(row: Int): Int
  def sumInCol(col: Int): Int

  def randomize
}
